synapse/util/patch_inline_callbacks.py

```diff
@@ -174,7 +174,9 @@ def check_yield_points_inner(
                         )
                     )
                     changes.append(err)
-                return getattr(e, "value", None)
+                # The `StopIteration` or `_DefGen_Return` contains the return value from the
+                # generator.
+                return cast(T, e.value)
```

Comment:
Both of the exception types we catch here have a `value` attribute:
```py
class StopIteration(Exception):
    value: Any

class _DefGen_Return(BaseException):
    def __init__(self, value: object) -> None:
        self.value = value
```

Since the exception comes from the generator `gen`, which is supposedly a `Generator[..., ..., T]`, the `StopIteration` ought to contain a `T`.
I can't say the same for `_DefGen_Return` because I have no idea what its purpose is. I'm guessing it works the same way.

Reply:
> Both of the exception types we catch here have a `value` attribute:
> 
> ```python
> class StopIteration(Exception):
>     value: Any
> 
> class _DefGen_Return(BaseException):
>     def __init__(self, value: object) -> None:
>         self.value = value
> ```
> 
> Since the exception comes from the generator `gen`, which is supposedly a `Generator[..., ..., T]`, the `StopIteration` ought to contain a `T`. I can't say the same for `_DefGen_Return` because I have no idea what its purpose is. I'm guessing it works the same way.

Seems plausible: https://github.com/twisted/twisted/blob/5c24e99e671c4082a1ddc8dbeb869402294bd0dc/src/twisted/internet/defer.py#L1508-L1521